File: Data/HighScoreManager.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.IO;

namespace Shortcutsy.Data;

/// <summary>
/// Represents a single high score entry.
/// </summary>
public class HighScoreEntry
{
    public int Score { get; set; }
    public int Level { get; set; }
    public string Date { get; set; } = "";
}

/// <summary>
/// Manages high score persistence, storing best score per level.
/// </summary>
public class HighScoreManager
{
    private const string FileName = "highscores.json";
    private Dictionary<int, int> _levelScores = new();
    private readonly string _filePath;
    private readonly Func<string, bool>? _fileExists;
    private readonly Func<string, string>? _fileReader;
    private readonly Action<string>? _fileWriter;

    /// <summary>
    /// Creates a HighScoreManager with default file-based persistence.
    /// </summary>
    public HighScoreManager() : this(null, null, null)
    {
    }

    /// <summary>
    /// Creates a HighScoreManager with custom file handling (for testing).
    /// </summary>
    public HighScoreManager(Func<string, string>? fileReader, Action<string>? fileWriter, Func<string, bool>? fileExists)
    {
        _filePath = Path.Combine(AppContext.BaseDirectory ?? "", FileName);
        _fileReader = fileReader;
        _fileWriter = fileWriter;
        _fileExists = fileExists ?? File.Exists;
    }

    /// <summary>
    /// Loads high scores from disk.
    /// </summary>
    public void Load()
    {
        try
        {
            if (_fileExists?.Invoke(_filePath) ?? File.Exists(_filePath))
            {
                string json = _fileReader?.Invoke(_filePath) ?? File.ReadAllText(_filePath);
                _levelScores = JsonSerializer.Deserialize<Dictionary<int, int>>(json) ?? new();
            }
        }
        catch
        {
            _levelScores = new();
        }
    }

    /// <summary>
    /// Saves high scores to disk.
    /// </summary>
    public void Save()
    {
        try
        {
            string json = JsonSerializer.Serialize(_levelScores, new JsonSerializerOptions { WriteIndented = true });
            if (_fileWriter != null)
            {
                _fileWriter(json);
            }
            else
            {
                File.WriteAllText(_filePath, json);
            }
        }
        catch { }
    }

    /// <summary>
    /// Adds a new score for a level. Returns true if it's a new record for that level.
    /// </summary>
    public bool AddScore(int score, int level)
    {
        bool isNewRecord = false;
        
        if (!_levelScores.ContainsKey(level) || score > _levelScores[level])
        {
            _levelScores[level] = score;
            isNewRecord = true;
        }
        
        Save();
        return isNewRecord;
    }

    /// <summary>
    /// Gets the high score for a specific level.
    /// </summary>
    public int GetScoreForLevel(int level)
    {
        return _levelScores.ContainsKey(level) ? _levelScores[level] : 0;
    }

    /// <summary>
    /// Gets all stored high scores as a list.
    /// </summary>
    public List<HighScoreEntry> GetScores()
    {
        return _levelScores.Select(kv => new HighScoreEntry { Level = kv.Key, Score = kv.Value, Date = "" })
            .OrderByDescending(s => s.Score)
            .ToList();
    }

    /// <summary>
    /// Gets all level scores as dictionary.
    /// </summary>
    public Dictionary<int, int> GetAllScores()
    {
        return new Dictionary<int, int>(_levelScores);
    }

    /// <summary>
    /// Clears all high scores (for testing).
    /// </summary>
    public void Clear()
    {
        _levelScores.Clear();
    }
}

File: Shortcutsy.Web/src/highScoreManager.ts
/**
 * HighScoreManager - uses localStorage instead of file I/O
 */

const STORAGE_KEY = 'shortcutsy_highscores';

export class HighScoreManager {
  private levelScores: Map<number, number> = new Map();

  load(): void {
    try {
      const json = localStorage.getItem(STORAGE_KEY);
      if (json) {
        const data = JSON.parse(json) as Record<string, number>;
        for (const [key, value] of Object.entries(data)) {
          this.levelScores.set(parseInt(key), value);
        }
      }
    } catch {
      this.levelScores = new Map();
    }
  }

  save(): void {
    try {
      const obj: Record<string, number> = {};
      for (const [key, value] of this.levelScores) {
        obj[key.toString()] = value;
      }
      localStorage.setItem(STORAGE_KEY, JSON.stringify(obj));
    } catch { /* ignore */ }
  }

  addScore(score: number, level: number): boolean {
    const current = this.levelScores.get(level) || 0;
    let isNewRecord = false;
    if (score > current) {
      this.levelScores.set(level, score);
      isNewRecord = true;
    }
    this.save();
    return isNewRecord;
  }

  getScoreForLevel(level: number): number {
    return this.levelScores.get(level) || 0;
  }

  getScores(): { score: number; level: number }[] {
    const entries: { score: number; level: number }[] = [];
    for (const [level, score] of this.levelScores) {
      entries.push({ score, level });
    }
    return entries.sort((a, b) => b.score - a.score);
  }

  getAllScores(): Map<number, number> {
    return new Map(this.levelScores);
  }

  clear(): void {
    this.levelScores.clear();
  }
}
